src/derive_utils.rs

```diff
@@ -31,3 +39,37 @@ impl<'a> From<&'a PyModule> for PyFunctionArguments<'a> {
         PyFunctionArguments::PyModule(module)
     }
 }
+
+/// Enum to abstract over the arguments of Python function wrappers.
+pub enum PyFunctionArgumentsBound<'a, 'py> {
+    Python(Python<'py>),
+    PyModule(&'a Bound<'py, PyModule>),
+}
```

Comment:
It is a fair bit of duplication, but I worry that trying to make it more generic will be very awkward. Maybe you can draft a refactored version for comparison?

Reply:
I have played around with it a bit. I have pushed it [here](https://github.com/PyO3/pyo3/compare/main...Icxolu:pyo3:pymodule-bound). This reduces the duplication at the expense of being a breaking change (but I believe that it should not really have a large impact, because users use `.into()` to convert a `&PyModule` into a `PyFunctionArguments`, which does still work). For the same reason we actually can't make the old API generic, because then these `.into()`  calls would be ambiguous.

I still believe the new ones should be tho.

Reply:
That's a lot less work than I feared! I'll leave it to other reviewers to decide which approach is preferred.

Reply:
I like that a lot! The intent is much clearer, we don't break compatibility and we can completely remove `PyFunctionArguments` together with the gil-refs in the future.

Reply:
@davidhewitt Just to be clear, did we want both of @Icxolu's suggested changes? I realised I changed these `PyCFunction` methods independently on this PR.

I've created another branch locally that builds on @Icxolu's changes which I can push over this if that's what we want.